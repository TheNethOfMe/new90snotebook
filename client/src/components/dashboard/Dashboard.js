import React, { Component } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import {
  getCurrentProfile,
  populateProfileFromStorage
} from "../../actions/profileActions";
import Spinner from "../common/Spinner";
import Notification from "./Notifications";

export class Dashboard extends Component {
  componentDidMount() {
    const storeString = window.localStorage.getItem("My90sNBStore");
    if (!storeString || !JSON.parse(storeString).hasOwnProperty("profile")) {
      this.props.getCurrentProfile();
    } else {
      const storeObject = JSON.parse(storeString);
      this.props.populateProfileFromStorage(storeObject.profile);
    }
  }
  render() {
    const { profile, loading } = this.props.profile;
    let dashboardContent;

    if (!profile || loading) {
      dashboardContent = <Spinner />;
    } else {
      dashboardContent = <Notification />;
    }
    return (
      <div className="dashbaord">
        <h2>Welcome</h2>
        {dashboardContent}
      </div>
    );
  }
}

Dashboard.propTypes = {
  getCurrentProfile: PropTypes.func.isRequired,
  populateProfileFromStorage: PropTypes.func.isRequired,
  profile: PropTypes.object.isRequired,
  auth: PropTypes.object.isRequired
};

const mapStateToProps = state => ({
  profile: state.profile,
  auth: state.auth
});

export default connect(
  mapStateToProps,
  { getCurrentProfile, populateProfileFromStorage }
)(Dashboard);
