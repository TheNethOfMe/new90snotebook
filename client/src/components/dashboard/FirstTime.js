import React, { Component } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import ProfileForm from "../settings/ProfileForm";
import { createNewProfile } from "../../actions/profileActions";

export class FirstTime extends Component {
  constructor(props) {
    super(props);
    this.state = {
      errors: {}
    };
  }
  componentWillReceiveProps(nextProps) {
    if (nextProps.errors) {
      this.setState({ errors: nextProps.errors });
    }
  }
  createProfile = newProfile => {
    this.props.createNewProfile(newProfile);
  };
  render() {
    const { errors } = this.state;

    return (
      <div>
        <h1>Profile Setup</h1>
        <ProfileForm
          profile={null}
          errors={errors}
          handleSubmit={this.createProfile}
        />
      </div>
    );
  }
}

FirstTime.propTypes = {
  errors: PropTypes.object.isRequired,
  createNewProfile: PropTypes.func.isRequired
};

const mapStateToProps = state => ({
  errors: state.errors
});

export default connect(
  mapStateToProps,
  { createNewProfile }
)(FirstTime);
