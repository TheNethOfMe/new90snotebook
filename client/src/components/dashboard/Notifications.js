import React, { Component } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { getAllNotifications } from "../../actions/notificationActions";
import Spinner from "../common/Spinner";

export class Notifications extends Component {
  componentDidMount() {
    this.props.getAllNotifications();
  }
  render() {
    const { notifications, loading } = this.props.notifications;
    let componentContent;

    if (!notifications || loading) {
      componentContent = <Spinner />;
    } else if (notifications.length === 0) {
      componentContent = <p>There are no notifications to display</p>;
    } else {
      componentContent = (
        <div>
          {notifications.map(note => (
            <p key={note._id}>{note.message}</p>
          ))}
        </div>
      );
    }
    return <div>{componentContent}</div>;
  }
}

Notifications.propTypes = {
  getAllNotifications: PropTypes.func.isRequired,
  notifications: PropTypes.object.isRequired
};

const mapStateToProps = state => ({
  notifications: state.notifications
});

export default connect(
  mapStateToProps,
  { getAllNotifications }
)(Notifications);
