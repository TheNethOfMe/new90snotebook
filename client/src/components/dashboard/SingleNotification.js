import React, { Component } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { deleteNotification } from "../../actions/notificationActions";

export class SingleNotification extends Component {
  handleDelete = e => {
    this.props.deleteNotification(this.props.id);
  };
  render() {
    return (
      <div className="notification__panel">
        <div> {this.props.message}</div>
        {this.props.closable && (
          <button onClick={this.handleDelete}>dismiss</button>
        )}
      </div>
    );
  }
}

SingleNotification.propTypes = {
  message: PropTypes.string.isRequired,
  closable: PropTypes.bool.isRequired,
  deleteNotification: PropTypes.func.isRequired
};

export default connect(
  null,
  { deleteNotification }
)(SingleNotification);
