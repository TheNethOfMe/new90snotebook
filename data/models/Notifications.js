const mongoose = require("mongoose");
const Schema = mongoose.Schema;

const NotificationSchema = new Schema({
  notificationFor: {
    type: Schema.Types.ObjectId,
    ref: "users"
  },
  message: {
    type: String
  },
  createdAt: {
    type: Date,
    default: Date.now
  }
});

module.exports = Notification = mongoose.model(
  "notification",
  NotificationSchema
);
