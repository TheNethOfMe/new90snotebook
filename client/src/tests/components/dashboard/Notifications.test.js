import React from "react";
import { shallow } from "enzyme";
import { Notifications } from "../../../components/dashboard/Notifications";
import { fakeNotification } from "../../dummyData";

let wrapper, getAllNotifications;

beforeEach(() => {
  getAllNotifications = jest.fn();
});

test("should render component and call getAllNotifications", () => {
  const notifications = {
    notifications: [fakeNotification],
    loading: false
  };
  wrapper = shallow(
    <Notifications
      notifications={notifications}
      getAllNotifications={getAllNotifications}
    />
  );
  expect(wrapper).toMatchSnapshot();
  expect(getAllNotifications).toHaveBeenCalled();
});

test("should render component with notifications", () => {
  const notifications = {
    notifications: [fakeNotification],
    loading: false
  };
  wrapper = shallow(
    <Notifications
      notifications={notifications}
      getAllNotifications={getAllNotifications}
    />
  );
  expect(
    wrapper
      .find("p")
      .at(0)
      .text()
  ).toEqual(fakeNotification.message);
});

test("should render component without notifications", () => {
  const notifications = {
    notifications: [],
    loading: false
  };
  wrapper = shallow(
    <Notifications
      notifications={notifications}
      getAllNotifications={getAllNotifications}
    />
  );
  expect(
    wrapper
      .find("p")
      .at(0)
      .text()
  ).toEqual("There are no notifications to display");
});
