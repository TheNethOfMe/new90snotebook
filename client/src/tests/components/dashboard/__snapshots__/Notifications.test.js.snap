// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render component and call getAllNotifications 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Notifications
    getAllNotifications={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    notifications={
      Object {
        "loading": false,
        "notifications": Array [
          Object {
            "id": 56,
            "message": "This is a notification",
            "notificationFor": "TESTID",
          },
        ],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div>
        <p>
          This is a notification
        </p>
      </div>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <p>
            This is a notification
          </p>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "This is a notification",
          },
          "ref": null,
          "rendered": "This is a notification",
          "type": "p",
        },
      ],
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div>
          <p>
            This is a notification
          </p>
        </div>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <p>
              This is a notification
            </p>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "This is a notification",
            },
            "ref": null,
            "rendered": "This is a notification",
            "type": "p",
          },
        ],
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
